### PR TITLE
Implemented missing metrics for GC detection

### PR DESCRIPTION
**Pull request by katarina-glorija**
LCOM3 - Lack of cohesion based on https://core.ac.uk/download/pdf/231161496.pdf
LCOM4 - Lack of cohesion based on http://www.isys.uni-klu.ac.at/PDF/1995-0043-MHBM.pdf
DIT -  Depth inheritance hierarchy based on https://objectscriptquality.com/docs/metrics/depth-inheritance-tree
DCC -  Direct class coupling based on http://www.theijes.com/papers/v5-i6/C05014019.pdf

### CODE REVIEW

File: CodeModel.Tests/Unit/MetricCalculationTests.cs [hunk 1]
```diff
@@ -69,6 +69,32 @@ public void Calculates_lack_of_cohesion()
...
+        public void Calculates_lack_of_cohesion_3()
+        {
+            CodeModelFactory factory = new CodeModelFactory();
+
+            List<CaDETClass> classes = factory.CreateProject(_testDataFactory.GetCohesionClasses()).Classes;
+
+            var dateRange = classes.Find(c => c.Name.Equals("DateRange"));
+            var doctor = classes.Find(c => c.Name.Equals("Doctor"));
+            dateRange.Metrics[CaDETMetric.LCOM3].ShouldBe(0);
+            doctor.Metrics[CaDETMetric.LCOM3].ShouldBe(1.125);
```
Comment (Luburic) [inline CodeModel.Tests/Unit/MetricCalculationTests.cs:82]: Did you take this test data and calculated values from the paper? Ideally we would have a test suite that matches the examples provided by the authors.

File: CodeModel.Tests/DataFactories/CodeFactory.cs [hunk 2]
```diff
@@ -767,6 +767,48 @@ IEnumerator IEnumerable.GetEnumerator()
             };
         }
 
+        public IEnumerable<string> GetCouplingClasses()
```
Comment (Luburic) [inline CodeModel.Tests/DataFactories/CodeFactory.cs:770]: This class needs to die. We'll schedule a meeting to plot its murder.

File: CodeModel/CaDETModel/CodeItems/CaDETClass.cs [hunk 3]
```diff
@@ -45,6 +45,20 @@ public List<CaDETClass> GetMethodLinkedVariableTypes()
...
+        {
+            List<CaDETParameter> parameters = Members.SelectMany(m => m.Params).ToList();
+            return parameters.Select(p => p.Type)
+                   .Where(v => v.LinkedTypes != null)
+                   .SelectMany(v => v.LinkedTypes).ToList();
+        }
+
+        public List<CaDETClass> GetMethodInvocationsTypes()
+        {
+            List<CaDETMember> invokedMethods = Members.SelectMany(m => m.InvokedMethods).Where(m => m.Type is CaDETMemberType.Method).ToList();
```
Comment (Luburic) [inline CodeModel/CaDETModel/CodeItems/CaDETClass.cs:58]: Invoked methods are already only methods (which is why we don't call them invoked members).

File: CodeModel/CodeParsers/CSharp/CaDETClassMetricCalculator.cs [hunk 4]
```diff
@@ -69,6 +74,55 @@ private double GetLackOfCohesionOfMethods(CaDETClass parsedClass)
...
+                methodFieldAccess += CountOwnFieldAndAccessorAccessed(parsedClass, method);
+            }
+            return Math.Round((numberOfMethods - methodFieldAccess/numberOfAttributes) / (numberOfMethods - 1), 3);
+        }
+
+        /// <summary>
+        /// LCOM - Lack of Cohesion of Methods
+        /// DOI: 10.1145/2723742.2723753
+        /// </summary>
+        private double GetLackOfCohesionOfMethods4(CaDETClass parsedClass)
```
Comment (Luburic) [inline CodeModel/CodeParsers/CSharp/CaDETClassMetricCalculator.cs:101]: Given the complexity of this calculation we need a lot of test data. Search online if the papers don't give enough examples and have 5+ test cases for each metric.

**Review by Luburic — CHANGES_REQUESTED** [review]
A few minor changes, plus a task to enhance the test suite with (many) more test cases gathered from online resources. Given the nature of metrics and the issues that arise from their (mis)implementation, we need to significantly expand our test suite.

File: CodeModel.Tests/Unit/MetricCalculationTests.cs (hunk 1 above)
Comment (katarina-glorija) [reply CodeModel.Tests/Unit/MetricCalculationTests.cs:82]: I calculated them manually. Unfortunately, the paper doesn't provide any example.

Comment (Luburic) [reply CodeModel.Tests/Unit/MetricCalculationTests.cs:82]: Then look to the web for inspiration.

**Review by Luburic — APPROVED** [review]
Looks good.

**Pull request merged**